### PR TITLE
Correlations: Allow creating correlations for provisioned data sources

### LINKED ISSUE #71110: Glue: Allow adding correlations for read-only data sources
__Background__

Design doc: https://docs.google.com/document/d/1BjJ1GDxY3BjFzF9ZZKe19gN7foyLDZa05k2U1YSlPyM/edit
Proposal: 1

__Requirements__

1. Remove validation code that prevents adding a correlation to a read-only data source (e.g. [source](https://github.com/grafana/grafana/blob/v10.0.1/pkg/services/correlations/database.go#L35) but applicable to update and delete as well)
2. Keep automatic removals when a data source is removed (this requires extra logic when a provisioned data source is re-created, see Technical Details below)
3. Provisioned correlations should be read-only. This requires changing how read-only is read in the frontend, it will require returning the `provisioned` property in the API (e.g. [source](https://github.com/grafana/grafana/blob/v10.0.1/public/app/features/correlations/CorrelationsPage.tsx#L33))
4. Update documentation (API and User docs)

Post-merge testing:
- Test in Cloud that renaming a stack keeps correlations intact 

__Technical details__

- Add `provisioned` column to database (alternatively, create a separate table like Dashboards and Alerts)
- Manually track what data sources are re-created during provisioning to clean-up unused resources

<details><summary>Provisioning algorithm</summary>

Check the design doc for more details. 

https://github.com/grafana/grafana/blob/v10.0.1/pkg/services/provisioning/datasources/datasources.go#L60 

1. Skip firing an event when the data source is deleted. Ideally: return a list of events when `dc.deleteDatasources` is called in provisioning, filter out data sources that got re-created, and fire events again to perform the cleanup.
2. Before provisioning data sources: remove all correlations where column `provisioned=true`
3. Create provisioned correlations as normal

</details>

_Suggested task split_

- [ ] Requirements: 2, 3. Add the `provisioning` column and change the provisioning logic. This can be tested by changing `provisioned` column in the db
- [ ] Requirements: 1, 4. Remove API validation, update frontend code, update documentation

### PR DESCRIPTION
**Pull request by ifrost**
Fixes https://github.com/grafana/grafana/issues/71110

This is to allow creating correlations for provisioned data sources. Please check https://github.com/grafana/grafana/issues/71110 for more details and design doc.

Key changes:

* Clean-up logic was adjusted to ensure correlations are removed when a data source is removed, taking into account a scenario when a data source is re-created during provisioning and no data loss occurs
* Previously, correlations attached to a read-only data source were marked as read-only. Now, only provisioned correlations are read-only

Notes on changes vs https://github.com/grafana/grafana/pull/71984:
* I needed to take multiple provisioning files into account, and decided to split provisioning of data sources and correlations into separate functions - should be a bit cleaner now
* As discussed in https://github.com/grafana/grafana/pull/71984#discussion_r1270515538 I moved the logic up, hopefully it's simpler to follow. With the split of provisioning data sources first there's also no need to have additional handling of the UID 
* We also discussed alternative approach: to generate a list of commands first to determine which data sources should be deleted or not. That was actually more complex to implement as we would need another level of abstraction for commands - service commands would need to be wrapper into "provisioning commands" because there's some additional logic around them (like reading the database to get the data source uid). Instead there's a simple map created at the beginning of the provisioning to indicate which data source will be deleted after provisioning finishes.

### CODE REVIEW

File: pkg/services/correlations/models.go
Comment (marefr) [inline pkg/services/correlations/models.go:12]: Nit. Would suggest using the errutil package for declaring errors which enable you to simplify the api.go error handling, see https://github.com/grafana/grafana/blob/main/contribute/backend/errors.md

No blocker for this PR

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (marefr) [inline pkg/services/sqlstore/migrations/correlations_mig.go:64]: Not sure this would work if the table already includes rows?

File: pkg/services/datasources/models.go
Comment (marefr) [inline pkg/services/datasources/models.go:157]: Not sure this will work. What if there are other consumers besides correlations relying on these events being published? Correlations delete handler somehow needs to make the decision to handle or not - can it check if the datasource is read only and skip it as an example?

Is there a problem correlations being deleted and re-created? Would simplify the code to separate/break datasource and correlation provisioning in two distinct packages/provisioners

**Review by marefr — CHANGES_REQUESTED** [review]
Looks good in general, but come questions/comments

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (ifrost) [reply pkg/services/sqlstore/migrations/correlations_mig.go:64]: Yes, it will not correctly mark existing records. Existing correlations may be created manually or be provisioned. However, I don't know if there's a way to figure out which ones are actually provisioned or not (there's no indication if the source data source is actually provisioned or not).

The idea to approach it is treat all existing records as not-provisioned, meaning that **even if a correlation was provisioned, it will be duplicated upon restart, marked as provisioned and the user would need to manually remove non-provisioned one. The duplication may happen only once, with no more duplicates piling up because: during provisioning all previously provisioned correlations are removed** (I should've added it to the description and changelog notes).

We do not expect a lot of problems related to duplicated entries. It's not possible to provision correlations in Cloud so no Cloud customers are affected. It's still early preview / opt-in features so we expect very limited impact for on-prem customers. 

However, alternatively we can:
- introduce some de-dup logic (e.g. on load) and remove duplicates based on equality of each column, or 
- add API and manual way to trigger de-duplication, or
- add a background clean-up process.

Adding any of the above however, may mean we need to maintain this code for long time, and I wasn't sure it worth adding if it has very low impact.

File: pkg/services/datasources/models.go
Comment (ifrost) [reply pkg/services/datasources/models.go:157]: > What if there are other consumers besides correlations relying on these events being published? 

The idea is that this is set only during provisioning and only when deleting data sources that will be re-created. New resources don't need to worry about this - the `DataSourceDeleted` won't be published at all so the handler won't be triggered. This happens only for data sources that are not truly deleted, but re-created. For example it is used in Cloud to rename a stack.

> Correlations delete handler somehow needs to make the decision to handle or not (...)

That would mean that every new consumer would need to handle it. There's a problem that we only know if a data source was read-only, not if it was provisioned. We could pass this information with `DeleteDataSource` event instead of guarding this with `SkipPublish`. However in our case, it's not a problem to have different logic during provisioning or not - it's that we want to perform clean-up only if the data source is actually deleted, not-recreated.

> Is there a problem correlations being deleted and re-created?

The main problem is that when a data source is removed we want to remove all created correlations (provisioned or done manually). However in Cloud, when the stack is renamed, a data source is removed and created once again through provisioning. Without `SkipPublish` all user-created correlations would be removed.

> Would simplify the code to separate/break datasource and correlation provisioning in two distinct packages/provisioners

You mean to have separate handlers for deletion happening during provisioning or not? In theory, as mentioned above, instead of `SkipPublish` we could always publish the event, but basically pass a boolean indicating if that happened during provisioning or not and let the handler decide if only provisioned resources should be removed or not. But then we end up in a similar problem as described above, but the other way round. If a data source is actually removed during provisioning, we also want to remove user-created correlations. 


I agree it's far from ideal. The main use case we want to cover is renaming the stack in Cloud. If that wasn't handled by deleting and creating datasources we could argue that if a data source is deleted in provisioning, it should have all resources removed even if it's re-created. But we know it's not fully true, and will create problems for Cloud customers.

Comment (ifrost) [reply pkg/services/datasources/models.go:157]: @marefr - does it make sense? Do you think we should take a different approach here?

Comment (marefr) [reply pkg/services/datasources/models.go:157]: Sorry for slow response. I'll try to come back with a response shortly

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (marefr) [reply pkg/services/sqlstore/migrations/correlations_mig.go:64]: Not sure you understood exactly what I meant. Historically we haven't been able to add a new column with nullable false and a default value in case there's been existing data in the table. Not sure if this applies to mysql or postgres, but maybe not sqlite.

For example for uid column added to the datasource table:
https://github.com/grafana/grafana/blob/364ca3daa0eab69bcec9f2b70cb08ee8b69e5ab3/pkg/services/sqlstore/migrations/datasource_mig.go#L117-L129

File: pkg/services/datasources/models.go
Comment (marefr) [reply pkg/services/datasources/models.go:157]: Got it I see the challenge now. Then I probably don't see any other alternative than your suggested approach or stop caring that datasources are deleted as dashboards does it.

> Another reason is that DeleteDataSource event was added for Correlation in https://github.com/grafana/grafana/pull/51630, so exiting resources do not use it, and AFAIK there's no other resources using it.

If someone starts using it then it might be an unexpected behavior that it's not triggered on provisioned datasources and it's very much hidden that this behavior exists.

**Comment by ifrost** [comment]
> I did have some trouble with the deleteDatasources

How did you remove the data source? Through the UI or API? Do you have reproduction steps?

File: pkg/services/provisioning/datasources/datasources.go
Comment (ifrost) [reply pkg/services/provisioning/datasources/datasources.go:59]: You mean to rename the prop? 🤔 Not sure if `oneWayDeletionList` makes it clearer what it does. I guess `ultimatelyDeleted` is also confusing. How about `not/recreatedAfterDeletion`? :D

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (ifrost) [reply pkg/services/sqlstore/migrations/correlations_mig.go:64]: Oh I see. I wasn't aware there might be problem with it. I'll double-check, test it and update migration if needed.

File: pkg/services/datasources/models.go
Comment (ifrost) [reply pkg/services/datasources/models.go:157]: > If someone starts using it then it might be an unexpected behavior that it's not triggered on provisioned datasources and it's very much hidden that this behavior exists.

The skipping is applied only in one case: when someone deletes and creates a data source with the same UID during single provisioning process. If the user deletes the data source by adding it to `deleteDatasources` list without creating a data source with the same uid - it will be deleted as normal.

I agree it's not clear this behavior exists. However, without it we end up with another unexpected behavior: "You can change your data source name by deleting and re-adding it, but... you will lose all resources related to your data source". 

I guess a better solution could be if we had an option in provisioning to rename a data source. Maybe that's the way to follow-up on this? With something like:

```
renameDatasources:
  - previous: Previous Name
     new: New Name
     uid: UID
```

We could use it in the Cloud to rename stacks and get rid of that event skipping.

**Comment by ifrost** [comment]
> One thing I did notice though, it copied the gdev provisioned correlations, they now show up twice. Once as readonly, once as not.

Yeah that's a drawback. In theory, we could add some logic to check for duplicates during provisioning (and mark something as provisioned instead of adding a new record). My thinking was that should be a rare case anyway, but I'lll have a look how it'd look like if we had that de-dup logic in place.

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (ifrost) [reply pkg/services/sqlstore/migrations/correlations_mig.go:64]: @marefr I've checked Postgres (15.4) and MySQL (8.0.32) and everything seemed to work fine.  I tested it with this scenario:

* create a clean database in main, 
* add some records in `correlation` table, 
* switch the this PR branch
* restart the service to run migration

In both cases the migration worked as expected and the default value was set up in all existing records correctly.

I also tried running the following SQL statements in Postgres and MySQL:

```sql
create table test (foo varchar(50));
insert into test(foo) values ('foo1');
insert into test(foo) values ('foo2');
alter table test add column bar varchar(50) not null default 'bar';
```

The last statement (representing the expected statement of the migration) passed and the default value was set up.

I also found this conversation: https://github.com/grafana/grafana/pull/23585#discussion_r408764230, but I struggle to figure out the case when the migration would not work as expected. 

Do you know what am I missing here? 🤔 Was this issue related to a specific db version? Or maybe was related to creating a column with unique values? I can also change `nullable=true` just in case as it should be prevented in the code anyway, but would be good to understand it better.

File: pkg/services/correlations/models.go
Comment (ifrost) [reply pkg/services/correlations/models.go]: Thanks! I've created a separate task for it to avoid to not add too much to this PR: https://github.com/grafana/grafana/issues/74523

File: pkg/services/correlations/api.go
Comment (ifrost) [reply pkg/services/correlations/api.go]: 👍 Updated

File: pkg/services/provisioning/datasources/datasources.go
Comment (ifrost) [reply pkg/services/provisioning/datasources/datasources.go]: Changed it to `willExistAfterProvisioning` :D what do you think?

**Comment by ifrost** [comment]
Reopening it for review after a few changes. @marefr, @gelicia - could you have another look? 

@gelicia I've updated replies to your comments in https://github.com/grafana/grafana/pull/73737#issuecomment-1700938556 and https://github.com/grafana/grafana/pull/73737#issuecomment-1702632499 with more details.

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (Elfo404) [inline pkg/services/sqlstore/migrations/correlations_mig.go:63]: shouldn't this be `correlationsV2` ?

```suggestion
	mg.AddMigration("add provisioning column", NewAddColumnMigration(correlationsV2, &Column{
```

Comment (ifrost) [reply pkg/services/sqlstore/migrations/correlations_mig.go:63]: Thanks for spotting this, I forgot to update after merging from `main`.

File: pkg/services/provisioning/datasources/config_reader_test.go
Comment (Elfo404) [inline pkg/services/provisioning/datasources/config_reader_test.go:252]: should we also test the new non-duplication logic?

**Review by Elfo404 — COMMENTED** [review]
Just 2 nits. otherwise LGTM. 

I didn't have any trouble applying the migration on SQLite, not approving as i'd wait for @gelicia or @marefr as they probably had a deeper look at this

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (marefr) [reply pkg/services/sqlstore/migrations/correlations_mig.go:64]: Hmm maybe it was when using older mysql version then

File: pkg/services/provisioning/datasources/config_reader_test.go
Comment (ifrost) [reply pkg/services/provisioning/datasources/config_reader_test.go:252]: The provisioning test uses mocked data base so I'm not sure if there's a way to test it e2e. I've added a test at the service level to check if the method works as expected here https://github.com/grafana/grafana/pull/73737/files#diff-c64b433d85c39bd84c2b456f3da5dcb22c913e5ede528ab6b38c478b99621f2a.

File: pkg/services/correlations/models.go
Comment (ifrost) [reply pkg/services/correlations/models.go]: Good catch. Removed in c7b6e37bf157f9ea8fac0fb82d2ef748af2923be

File: pkg/services/sqlstore/migrations/correlations_mig.go
Comment (marefr) [reply pkg/services/sqlstore/migrations/correlations_mig.go:64]: I think this should go through a Hosted Grafana Feature Readiness Review in case you haven't considered that to get some additional operator eyes on this.

File: pkg/services/datasources/models.go
Comment (marefr) [reply pkg/services/datasources/models.go:157]: > I guess a better solution could be if we had an option in provisioning to rename a data source. Maybe that's the way to follow-up on this? With something like:

Sorry probably confused again. Unless you rename a datasource UID (which shouldn't happen) any correlations created referencing this datasource uid should continue to function given that you don't delete correlations when a datasource is deleted. So maybe what I'm suggesting is to not delete correlations when datasource is deleted. Doesn't matter I guess since we've already discussed this and you landed in this solution.

**Review by marefr — APPROVED** [review]
LGTM, but please note I haven't closely reviewed the last round of changes and I might not agree fully with the solution but it's fine I guess. I think this should go through a Hosted Grafana Feature Readiness Review in case you haven't considered that to get some additional operator eyes on this.

**Comment by ifrost** [comment]
Thanks a lot, @marefr! 🤝 

> (...) I might not agree fully with the solution (...)

How could we get a full agreement on the solution? What we would like to avoid is: making this feature available to customers, advertising it, and having to roll it back later.

> (...) I think this should go through a Hosted Grafana Feature Readiness Review (...)

Good point. @grafana/hosted-grafana-team we have discussed this feature in [the design doc](https://docs.google.com/document/d/1BjJ1GDxY3BjFzF9ZZKe19gN7foyLDZa05k2U1YSlPyM/edit). The implementation changed compared to the original proposal and I've documented the change in [Delivery section](https://docs.google.com/document/d/1BjJ1GDxY3BjFzF9ZZKe19gN7foyLDZa05k2U1YSlPyM/edit#heading=h.7z6utqntlet3).

We've gone through [FRR](https://docs.google.com/document/d/1-OTyd8KizDDxM9pR-cna1LB7fEcvu7XgXoP181GJIsQ/edit) for the entire feature in Feb and this change should not affect it. We expect almost no impact on the startup of Cloud instances as correlations are not provisioned in Cloud. During stack renames, we won't run any additional queries so we don't expect having any impact on this as well.

**Review by gelicia — APPROVED** [review]
Looks great! Tested the following:
1. created correlation in ui on provisioned datasource that is not deleted and restarted server. correlation still existed, as expected
2. created correlation in ui on provisioned datasource that is deleted and restarted server. correlation still existed, as expected
1. created correlation in provisioning on provisioned datasource that is not deleted and restarted server. correlation exists as read only, as expected
1. created correlation in provisioning on provisioned datasource that is deleted and restarted server. correlation exists as read only, as expected
1. existing readonly correlations not duplicated
1. created correlation in provisioning on editable, provisioned datasource that is not deleted. correlation still exists as readonly. clearing correlations table and restarting server still shows correlation in readonly. this is as expected, as we do not want changes to provisioned correlations.

**Comment by marefr** [comment]
> How could we get a full agreement on the solution?

I'm not sure we can that's why I approved :smile:

**Comment by ifrost** [comment]
> > How could we get a full agreement on the solution?
> 
> I'm not sure we can that's why I approved 😄

> So maybe what I'm suggesting is to not delete correlations when datasource is deleted. Doesn't matter I guess since we've already discussed this and you landed in this solution.

Thanks @marefr! Just to make sure we're on the same page. If I understand correctly we have two different options on the table that we don't have full agreement on:

1) Keep the logic to remove related correlations when a data source is removed
2) Do not delete correlations when a data source is removed

Or is there anything else we don't agree on?

Explore squad discussed option (2) in [the design doc](https://docs.google.com/document/d/1BjJ1GDxY3BjFzF9ZZKe19gN7foyLDZa05k2U1YSlPyM/edit#heading=h.z06bb7zfovhl) (Proposal 2) and we weighted current proposal over it, but I think it's worth re-considering.

To recap the pros and cons of both options:

1) 👍 Resources are cleaned up, 👍 no orphaned records, 👎 harder to maintain with extra logic around skipping delete events
2) 👍 Code is easier to maintain, 👍 more consistent with dashboards and alerts, 👎  orphaned records left in the database

Switching to option (2) later should not affect this functionality from the customer's point of view so what I suggest is:
* Move forward with this PR 
* Create a design doc to discuss how to move forward (I can imagine option 3 - adding an option to rename data sources during provisioning, e.g. `renameDataSources` section, and remove "skip event" logic + avoid having orphaned records at the same time)

I think it's important to at least have a consistent way of handling this in the future. Cleaning-up correlations were introduced before this PR in #51630, so we're not changing anything with this PR. We thought that other resources could follow suit with a similar logic but as we see it has caveats on its own. It means we currently have 2 different ways of handling related resources (e.g. correlations are removed, alerts are not), and we should look into having a consistent logic. 

What do you think?

**Comment by marefr** [comment]
> Move forward with this PR

SGTM :+1: 

> Create a design doc to discuss how to move forward (I can imagine option 3 - adding an option to rename data sources during provisioning, e.g. renameDataSources section, and remove "skip event" logic + avoid having orphaned records at the same time)

I'll leave that up to you. In the future I guess that we definitely want to remove orphaned records everywhere, but probably something the app platform team are thinking of.

**Pull request merged**
